Implement Changeover Delay for Hydronic Low Temperature Radiant Systems

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #8057 
 - This pull request implements the new feature for the variable and constant flow low temperature radiant systems in EnergyPlus that allows the user to lock out or prevent the system from switching immediately from heating to cooling or vice versa.  This changeover delay allows the user to lock the system out of making this switch for a period of time that is user defined via a new schedule input field.  This gives the user full flexibility in the implementation of this changeover delay.  Users who do not wish to prevent immediate changeover can simply leave this new field blank or provide a schedule that is always zero.

The new code may result in a few minor changes to some of the existing files for the inlet and outlet temperatures to the radiant system when the system is actually off.  This is because the modified code fixed a small issue where the system could be off but carryover a temperature from a previous time step.  Now, when things are off, the inlet and outlet temperatures should always go back to the loop temperature.  So, in effect, the changes are because a minor bug was fixed along the way.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [no transition needed, field added to the end of the current syntax] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
NFP-Design Document Draft

This commit includes the initial/draft version of the NFP for this work.

RKStrand committed:
Change made based on Edwin's review comment

Change the NFP to allow the delay changeover be a scheduled value.

RKStrand committed:
Merge branch 'develop' into 8057-LowTempRadSysChangeoverDelay

RKStrand committed:
Implementation of Changeover Delay

This commit includes the code needed to implement the changeover delay, the IDD modifications, and two new input files that exercise the new functionality.  Unit test and docs to follow in a follow-up commit.

RKStrand committed:
Merge branch 'develop' into 8057-LowTempRadSysChangeoverDelay

RKStrand committed:
Unit test and documentation

This commit provides two unit tests, documentation, an input file fix, and a minor code modification.  This should complete the work for this modification and it is ready for review now.

RKStrand committed:
Corrected errors in IDF

One IDF file is failing even though it is running fine locally on my machine.  This commit eliminates all of the warning messages in hopes that it will clear up this problem.

Myoldmopar commented:
@RKStrand there are some conflicts here, presumably conflicting with changes you made in the last PR.  Before I jump in to do final review, could you get those conflicts resolved and pushed up?

RKStrand commented:
@Myoldmopar Absolutely.  I will work on that ASAP and resolve the conflict.  It's almost certain to be a result of the other radiant thing that got merged yesterday.  So, I'm not at all surprised.

RKStrand committed:
Merge branch 'develop' into 8057-LowTempRadSysChangeoverDelay

RKStrand committed:
Corrections to merge issues

Merging in other recent radiant system work introduced a few errors.  This commit resolves the problems.

Myoldmopar commented:
@RKStrand it looks like the number of big mathdiffs has increased as this PR came along, to where we now have 4 files with big diffs.  You mention in the body of the original post that some diffs are expected, but there's nothing on here to verify the changes are as-expected.  Can you provide justification of the results changes?

RKStrand commented:
@Myoldmopar The diffs that are happening in those two of those files are related to a fix to the condensation control shut-down.  In the previous code, it wasn't good about changing the operating mode to "not operating" if it shutdown.  I needed to change that so that the changeover delay stuff would work properly and capture the actual operating mode.  So, this was a bug that I needed to fix to get the changeover algorithm to work properly.  It involved adding four lines of:

this->OperatingMode = NotOperating;

at lines 3625, 3785, 4758, and 4820.  Before those were added, the radiant system would have zero flow and come back with a non-sensical inlet and outlet temperature.  It should be the water temperature from the supply side.  Instead, it would be some other value (not even really sure how it got it).  Now, it is the supply side water temperature like it should be for both inlet and outlet conditions.  Again, those four fixes now trap the condensation issue correctly and turn things off as well as set the operating mode to not operating.

In HydrHeatCool and HydrISOHT, you see the big differences because both have condensation controls of some kind (so it has times when the system gets turned off due to condensation) and if you look at the diffs the diffs are in the inlet and outlet water temperatures.  Notice that the diffs in the inlet and outlet temperatures are identical and if you check out the actual csv files you will see that the inlet and outlet temperatures are now what they are supposed to be (10C) rather than some other value.  So, the diffs are actually justified based on the fix that I had to make and the output actually makes more sense now.

The other two files (CondFD) I thought were exhibiting the same thing, but actually neither has condensation controls.  They are exhibiting the same diffs though--same temperature change to inlet and outlet and no change in energy.  I just ran one of those files here locally and the results and the inlet and outlet temperatures look like they are supposed to (inlet temperature of 10C) so I'm inclined to think this is working properly.  It's possible that in the systematic setting of this->OperatingMode that I inadvertently fixed something else?  Not really sure but I can't imagine why the inlet temperatures for this case would have been anything other than 10C.  So, there is a diff here but it seems like the output from develop isn't right based on what I saw locally.

Does any of this help?

Myoldmopar commented:
It helps a lot.  I appreciate it.  I just needed to make sure the diffs were acceptable, and it sounds like they are.  I'll look over the changes.

Myoldmopar commented on doc/input-output-reference/src/overview/group-radiative-convective-units.tex:
```diff
@@ -1171,6 +1171,10 @@ \subsubsection{Inputs}\label{inputs-5-022}
 
+\paragraph{Field: Changeover Delay Time Period Schedule}\label{field-changeover-delay-time-period-schedule}
+
+This field defines a schedule which should be populated with values for the amount of time in hours that is required before this low temperature radiant system can switch between either heating or cooling and the opposite conditioning mode.  If this schedule value is set to 1 (hour) at a particular time and the system was previously in heating mode, then it will be locked out from switching into  ...
```
Ideally we would use the native LaTeX quotations: two back ticks to start the phrase, and then two apostrophes to close it:  ` ``my quote'' `.  It's not a critical thing here unless other things need to change.

RKStrand replied:
Oh snap.  Mike Witte called this out in a previous PR and since this was done before that other PR got reviewed...  If I need to fix something else, I'll snag this along the way.

Myoldmopar commented on doc/input-output-reference/src/overview/group-radiative-convective-units.tex:
```diff
@@ -1208,7 +1212,8 @@ \subsubsection{Inputs}\label{inputs-5-022}
       CalculateFromCircuitLength,   !- Number of Circuits
-      106.7;                   !- Circuit Length
+      106.7,                   !- Circuit Length
+      ChangeoverSchedule;      !- Changeover Delay Time Period Schedule
```
New input in the example, got it.

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -43990,7 +43990,7 @@ ZoneHVAC:LowTemperatureRadiant:VariableFlow,
         \memo surface (wall, ceiling, or floor). Controlled by varying the hot or chilled water
         \memo flow to the unit.
-        \min-fields 33
+        \min-fields 34
```
OK, so the new input is a required field on the end?

RKStrand replied:
Yes

Myoldmopar commented on idd/Energy+.idd.in:
```diff
@@ -44175,15 +44175,21 @@ ZoneHVAC:LowTemperatureRadiant:VariableFlow,
+   N16, \field Circuit Length
         \units m
         \default 106.7
+   A18; \field Changeover Delay Time Period Schedule
```
This doesn't explicitly say it is required, though you do have "minimum fields" incremented up one.  Perhaps we don't need to increment min-fields after all?

RKStrand replied:
@Myoldmopar Yeah, I'm still a bit fuzzy on that (you'd think I'd have it down after all these years).  Technically, it isn't required and the code just assumes no changeover delay if the schedule isn't there.  So, I guess it really isn't required and I should change the min-fields field?

Myoldmopar commented on tst/EnergyPlus/unit/LowTempRadiantSystem.unit.cc:
```diff
@@ -2469,6 +2469,164 @@ TEST_F(LowTempRadiantSystemTest, calculateRunningMeanAverageTemperatureTest)
+    thisRadSys.setOperatingModeBasedOnChangeoverDelay();
+    expectedResult = LowTempRadiantSystem::CoolingMode;
+    EXPECT_EQ(thisRadSys.OperatingMode, expectedResult);
+}
```
**Great** suite of tests here.

RKStrand replied:
Thanks! 😊

Myoldmopar commented on src/EnergyPlus/LowTempRadiantSystem.cc:
```diff
@@ -786,6 +784,18 @@ namespace LowTempRadiantSystem {
+            
+            thisRadSys.schedNameChangeoverDelay = Alphas(18);
+            if (lAlphaBlanks(18)) {
+                thisRadSys.schedPtrChangeoverDelay = 0;
```
Probably could've just relied on initializing the member variable to zero and not setting it to zero here, but it's OK.

RKStrand replied:
You are correct--could've been done as you suggested.  Still working out those F90 tendencies.  I can fix this if there are other things to change.

RKStrand replied:
Well, still "working it out" is right.  I actually already initialized it in the .hh. 😂 So, I had a hybrid.  Progress, right?  I'll get rid of the unnecessary assignment in line 790.

Myoldmopar commented on src/EnergyPlus/LowTempRadiantSystem.cc:
```diff
@@ -1367,6 +1389,12 @@ namespace LowTempRadiantSystem {
+                                thisHydrSys.OperatingMode,
+                                "System",
+                                "Average",
+                                thisHydrSys.Name);
```
Reporting an integer operating mode is a little fuzzy, but there is precedent, it's OK.

RKStrand replied:
Granted it's a bit weird, but it gives the user some idea what's going on.  That's why I changed cooling mode to -1.  So, if the output is less than 0, the user will know cooling took place.  It was also a help for verifying that the changeover delay was working and being interpreted inside the code properly.

Myoldmopar commented on src/EnergyPlus/LowTempRadiantSystem.cc:
```diff
@@ -2287,16 +2341,90 @@ namespace LowTempRadiantSystem {
+            // step level and will make decisions based on that.
+        this->lastDayOfSim  = DataGlobals::DayOfSim;
+        this->lastHourOfDay = DataGlobals::HourOfDay;
+        this->lastTimeStep  = DataGlobals::TimeStep;
```
This should be indented.

RKStrand replied:
You're right.  I totally missed that one.  Usually I'm pretty consistent with that, but clearly I didn't indent this properly here.

Myoldmopar commented on src/EnergyPlus/LowTempRadiantSystem.cc:
```diff
@@ -2287,16 +2341,90 @@ namespace LowTempRadiantSystem {
+            this->lastDayOfSim  = DataGlobals::DayOfSim - 1;
+            this->lastHourOfDay = int(DataGlobals::HoursInDay);
+            this->lastTimeStep  = DataGlobals::NumOfTimeStepInHour;
+        } else if (DataGlobals::BeginHourFlag) {
```
Can you confirm the nature of these blocks?  This component will be called multiple times during the same BeginDayFlag state.  And it will be called multiple times during BeginTimeStepFlag.  We've made fixes to other component models where we had to be very careful about resetting things.  If you can just confirm that this is fully thought through, that would be great.

RKStrand replied:
Yeah, it's a bit tricky and I had to think a bit about how to do that.  What allows this to work is that this routine doesn't get called at all if !FirstHVACIteration.  Without that, it would get called multiple times.  With that condition, it should only get called the absolute first time through.  So, I think the logic is correct unless I am missing something.  Does that help?  Or am I missing something?

Myoldmopar reviewed (commented):
The biggest concerns I have are the IDD min-fields change and the function that updates history based on different flags.  If you can comment on both of those, we'll push this closer to merging.

RKStrand commented:
@Myoldmopar Based on the answers above, let me know what you would like me to change and I'll take care of it as soon as I can.  I know the IO freeze deadline is very soon so I want to make sure this gets in.

Myoldmopar commented:
@mjwitte do you have any thoughts on the min-fields change here?  @RKStrand my suggestion would be to revert the min-fields locally and run a file in three combinations:
 - With this new field populated with a valid schedule name
 - With this new field in the IDF, but blank
 - WIth this new field completely missing (the object terminates after the previous field)

You could also try to break it by using an invalid schedule name, just for full posterity.

If all those work, then I would say going back to the previous min-fields is the best path forward.

mjwitte commented:
@Myoldmopar It doesn't really hurt for \min-fields to include this last field. But since the field is neither \required nor has a \default, \min-fields can safely stop short of this field (i.e. 33).

RKStrand commented:
@Myoldmopar I'll try your suggestions shortly and if it passes all the tests I'll fix this and the other minor issues above.  Once that's good, I'll commit and let CI do it's thing.  I'll keep you posted.

RKStrand committed:
Correction of Review Comments

A few minor changes based on review comments.

RKStrand commented:
@Myoldmopar Commit just got pushed.  All changes made as described above (unless I missed something).  I did all four tests that you recommended for both new input files.  Good thing you asked for the "posterity" test--it actually uncovered a bug in the new code!  That's been fixed and the case that has a bogus schedule does now give a warning message like it should.  All the other tests respond like they should with min-fields at 33.  So, that's what I committed.  Hopefully CI comes back clean.

RKStrand committed:
Merge branch 'develop' into 8057-LowTempRadSysChangeoverDelay

RKStrand commented:
Merged develop in so CI will run again.  Just to make sure things are all good.

Myoldmopar commented:
Audit diffs are trivial, and ERR diffs look to clean up some warnings.  Thanks for your assessment on the math diffs.  I believe this is good to go in.  Merging.  Thanks @RKStrand
